src/index.js

```diff
@@ -112,20 +128,105 @@ export const useReduxState = () => {
   const store = useContext(ReduxStoreContext);
   // redux state
   const state = store.getState();
+  // cache
+  const cacheRef = useRef({
+    proxy: new WeakMap(),
+    trapped: new WeakMap(),
+  });
   // proxyfied
-  const { proxyfiedState, lastProxyfied } = useProxyfied(state);
+  const proxyfied = createProxyfied(state, cacheRef.current);
+  // ref
+  const lastProxyfied = useRef(null);
+  useLayoutEffect(() => {
+    lastProxyfied.current = {
+      originalState: proxyfied.originalState,
+      affected: collectValuables(proxyfied.getAffected()),
+    };
+  });
+  // subscription
+  useEffect(() => {
+    const callback = () => {
+      const nextState = store.getState();
+      const changed = !proxyCompare(
+        lastProxyfied.current.originalState,
+        nextState,
+        lastProxyfied.current.affected,
+      );
+      drainDifference();
+      if (changed) {
+        lastProxyfied.current.originalState = nextState;
+        forceUpdate();
+      }
+    };
+    // run once in case the state is already changed
+    callback();
+    const unsubscribe = store.subscribe(callback);
+    return unsubscribe;
+  }, [store]); // eslint-disable-line react-hooks/exhaustive-deps
+  return proxyfied.trappedState;
+};
+
+export const useReduxSelectors = (selectorMap) => {
+  const forceUpdate = useForceUpdate();
+  // redux store
+  const store = useContext(ReduxStoreContext);
+  // redux state
+  const state = store.getState();
+  // keys
+  const keys = Object.keys(selectorMap);
+  // cache
+  const cacheRef = useRef({
+    selectors: new WeakMap(),
+  });
+  // mapped
+  const mapped = createMap(keys, (key) => {
+    const selector = selectorMap[key];
+    if (!cacheRef.current.selectors.has(selector)) {
+      cacheRef.current.selectors.set(selector, {
+        proxy: new WeakMap(),
+        trapped: new WeakMap(),
+        partialProxyfied: new WeakMap(),
+      });
+    }
+    const cache = cacheRef.current.selectors.get(selector);
+    if (cache.partialProxyfied.has(state)) {
+      const partialProxyfied = cache.partialProxyfied.get(state);
+      partialProxyfied.resetAffected();
+      return partialProxyfied;
+    }
+    const proxyfied = createProxyfied(state, cache);
+    const partialState = selector(proxyfied.trappedState);
+    const partialProxyfied = createProxyfied(partialState, cache);
```

Comment:
So you are recording access inside, and outside, in case you are returning some objects from selectors.
That's interesting, but we could make it better in terms of performance and reliability.

What I did before - go throught selectors result removing all the proxies from it. If proxy is used, then, by fact, you need object as a "whole".
It makes tracking more accurate, removing false negatives, and might improve perfomance, as long as you are no longer work with proxies inside a component.

See https://github.com/theKashey/memoize-state/blob/master/src/call.js#L29-L80

Reply:
So, is it like this?
```javascript
    const proxyfied = createProxyfied(state, cache);
    selector(proxyfied.trappedState); // just to record the usage
    const partialState = selector(state);
    const partialProxyfied = createProxyfied(partialState, cache);
```
assuming the purity of the selector.

Reply:
That would kill `reselect` if one is using it for memoization (habits), as long as variables in the first and the second call would be different.

It's more about
```js
    const proxyfied = createProxyfied(state, cache);
    return deproxify(selector(proxyfied.trappedState), trackResultUsed)
    // there would be "useAffected" keys, it the ones you have used, 
    // and "resultUsed" - the ones you have returned.
```

Simple example of "resultUsed" - 
```js
const selector = (state) => state.items.length ? state.items : [];
// access one key, return another.
```
- Default behaviour would track access to `.length` and thus __ignore__ any other change to the object. That's how `collectValuables` works - once you exposed your interest on variable internals - variable itself does not count.
- Your `partialProxyfied` would later track access to the items, if they would be _consumed_ in place, but if object would be passed to another object - probably not.
- `deproxify` would track `items` usage, and trigger update to any changes to it.

"Ideal" situation is to use your approach and patch `React.createElement` to _deproxyify_ and track a "whole objects" passed down as a props.

I have no idea how "your" approach would work with current `useEffect`, where you could pass an "object" to a child, and child could use it, and it would be tracked  - there are so many unknowns. And that's why `useLayoutEffect` is more predictable.

A year ago I've decided to scope all _decisions_ to `mapStateToProps` and not any proxy let out. Result is not so _granual_ as it could be, but more _reliable_. False positives a just small a performance problem, while false negatives are ☠️

Reply:
> That would kill reselect if one is using it for memoization (habits), as long as variables in the first and the second call would be different.

Oh, yes...

> go throught selectors result removing all the proxies from it. If proxy is used, then, by fact, you need object as a "whole".

If I understand this, it's going to be something like:
```
    const proxyfied = createProxyfied(state, cache);
    const partialState = deproxifyAll(selector(proxyfied.trappedState));
    const partialProxyfied = createShallowProxyfied(partialState, cache);
```

where `deproxifyAll` would deproxify recursively in an object, and createShallowProxyfied would only track an object direct properties (without collectValuables).

> useLayoutEffect is more predictable.

I don't understand why useLayoutEffect helps here. If only we don't use collectValuables like above, should be ok?

Reply:
There is no need of `createShallowProxyfied` of a `partialState` - it's "free".

`useLayoutEffect` could isolate tracking within one render. To be more concrete - it might be a good this to `seal` proxy after render, and `unseal` before. `.reset` does almost the same, but not stopping tracking for a proxied objects, you may "give" to children, while their access may lead to a false negatives.

```js
const items = state.items;
<Child items={items} />
// depends of items, update then it changed

Child = ({items})=>items.lenght;
// depends on items.lenght, it changed the parent! still tracked!
```

Reply:
So, I see that I don't yet understand various things. Please bear with me.

> There is no need of createShallowProxyfied of a partialState - it's "free".

The example:
```javascript
const Foo = ({ id }) => {
  const { bar } = useReduxSelectors({
    bar: useCallback(state => state[id], [id]),,
  });
  return <div>{bar.name}</div>
```
If `id=3`, the final affected is `.3.name` if I use use `collectValuables`, but I want it to be `.3`. How would you do that?

> it might be a good this to seal proxy after render, and unseal before.

Is this specific to `useReduxSelectors` or does it also apply to `useReduxState`?

Reply:
> If id=3, the final affected is .3.name if I use use collectValuables, but I want it to be .3. How would you do that?

you will find a `proxy` in a result and it would tell you _what it is_ - https://github.com/theKashey/memoize-state/blob/master/src/call.js#L29-L80

> Is this specific to useReduxSelectors or does it also apply to useReduxState?

If you will _deproxyfy_ result - then no further tracking is possible. What would also result to only `.3` got tracked, but imagine this code

```js
  const { bar } = useReduxSelectors({
    bar: useCallback(state => state[id].name ? state[id] : {name:'noname'}, [id]),
  });
  return <div>{bar.anotherField}</div>
``` 
`collectValueables` would report `.name` and `.anotherField`, with _further_ tracking enabled, and only `.name` if not.
So you have to scan _result_ of a selector to understand what was done inside.

Reply:
> https://github.com/theKashey/memoize-state/blob/master/src/call.js#L29-L80

Still reading this...

Reply:
Yeah, that might be not the best thing to read :)

Reply:
> collectValueables would report .name and .anotherField, with further tracking enabled, and only .name if not.

That's what I want to do. just track `.name`, not `.name` and `.anotherField`.
So, it looks like we are not talking about the same goal. Am I missing something here...

Reply:
I still think what I would like to do is:
```javascript
const proxyfied = createProxyfied(state, cache);
const partialState = deproxifyResult(selector(proxyfied.trappedState));
const partialProxyfied = createShallowProxyfied(partialState, cache);
```
Only the question is whether I'm on the right track with the requirement of respond-framework.

Reply:
So first you are checking “do I need to rerun a selector”, and later, if yes, “do I need to rerun render”?
It makes things a bit more complicated, do you actually need this secondary tracking?

Reply:
> So first you are checking “do I need to rerun a selector”, and later, if yes, “do I need to rerun render”?

Maybe a bit different.
1) check which part of state is used in a selector (=affected)
2) check if the selector result is used in render
3) trigger forceUpdate if any of affected of _used_ selectors is changed in a new state

> It makes things a bit more complicated, do you actually need this secondary tracking?

It may sound contradictory to you, but to me, if we don't have the secondary tracking, this hook doesn't provide any new tracking functionality and `useReduxState` is just enough.
Note that the secondary tracking is the requirement from respond-framework: https://github.com/dai-shi/reactive-react-redux/issues/16#issuecomment-478391434

Reply:
And I guess now I understand your point of view:
```javascript
const proxyfied = createProxyfied(state, cache);
const partialState = deproxifyResult(selector(proxyfied.trappedState));
```
and trigger forceUpdate if any of affected of all selectors is changed in a new state.
The point here is to hide proxy within selectors, and not in render.

Reply:
> trigger forceUpdate if any of affected of used selectors is changed in a new state

Nope :) selector is providing some __derived__ data from state, usually it's impossible to map `selected state` to the `real state`. They could be different. That's the idea of a selector - pick and transform (thus we need memoization), not only pick.

Reply:
Oh, I just noticed that my current implementation doesn't handle memoization well. I don't want to implement that so I should probably learn more about memoize-state and see if I can incorporate with it.

My previous question is still valid. I'd like to see how my scenario can be broken.

Reply:
- 👍  check which part of state is used in a selector (=affected) 
- 👍  check if the selector result is used in render
- 👎  trigger forceUpdate if any of affected of used selectors is changed in a new state
- 👍   trigger forceUpdate if any of affected of used selectors is changed in a __new selector value__.

You may track only what you put inside. So you will have to remember the last state value(to rerun selector), and the last selector result(to rerun render)

Reply:
Ooookay, I got it now. It's false positives, not false negatives, right? (My brain is only around false negatives...)

 Thanks for helping me out. It must be so natural to you, but I'm not yet used to the selectors world.

Reply:
Here's my plan:

1. check which part of state is used in a selector (=affected)
2. check if the selector result is used in render
3. when a state is updated, check any of affected of used selectors is changed in a new state, if so continue
4. rerun used selectors only if affected
5. trigger forceUpdate if any of results of used selectors is changed from the previous results

We might be able to skip step3 but if I'd keep it in case the selectorMap is big like having hundreds of items.

Reply:
Sounds good. Like you will have two filters - one before "`mapStateToProps`" and one after.